File: weather_service.py
import requests
from typing import Dict, Optional
from datetime import datetime

class WeatherService:
    """Weather information and intelligent analysis"""
    
    def __init__(self, api_key: str = None):
        """
        Initialize weather service
        
        Args:
            api_key: OpenWeatherMap API key (get free at openweathermap.org)
        """
        self.api_key = api_key
        self.base_url = "http://api.openweathermap.org/data/2.5"
        self.default_city = "New York"  # Can be customized
    
    def get_weather(self, city: str = None) -> Optional[Dict]:
        """
        Get current weather for a city
        
        Args:
            city: City name (uses default if not provided)
            
        Returns:
            Weather data dictionary or None if failed
        """
        if not self.api_key:
            return None
        
        if not city:
            city = self.default_city
        
        try:
            url = f"{self.base_url}/weather"
            params = {
                'q': city,
                'appid': self.api_key,
                'units': 'metric'  # Celsius
            }
            
            response = requests.get(url, params=params, timeout=5)
            
            if response.status_code == 200:
                data = response.json()
                return {
                    'city': data['name'],
                    'temperature': data['main']['temp'],
                    'feels_like': data['main']['feels_like'],
                    'humidity': data['main']['humidity'],
                    'description': data['weather'][0]['description'],
                    'wind_speed': data['wind']['speed'],
                    'icon': data['weather'][0]['icon']
                }
            return None
        except Exception as e:
            print(f"Error fetching weather: {e}")
            return None
    
    def get_forecast(self, city: str = None, days: int = 3) -> Optional[list]:
        """Get weather forecast for upcoming days"""
        if not self.api_key:
            return None
        
        if not city:
            city = self.default_city
        
        try:
            url = f"{self.base_url}/forecast"
            params = {
                'q': city,
                'appid': self.api_key,
                'units': 'metric',
                'cnt': days * 8  # 8 forecasts per day (3-hour intervals)
            }
            
            response = requests.get(url, params=params, timeout=5)
            
            if response.status_code == 200:
                data = response.json()
                forecasts = []
                
                for item in data['list']:
                    forecasts.append({
                        'time': datetime.fromtimestamp(item['dt']),
                        'temperature': item['main']['temp'],
                        'description': item['weather'][0]['description'],
                        'humidity': item['main']['humidity']
                    })
                
                return forecasts
            return None
        except Exception as e:
            print(f"Error fetching forecast: {e}")
            return None
    
    def get_weather_emoji(self, description: str) -> str:
        """Get emoji for weather condition"""
        description_lower = description.lower()
        
        if 'clear' in description_lower:
            return '☀️'
        elif 'cloud' in description_lower:
            return '☁️'
        elif 'rain' in description_lower:
            return '🌧️'
        elif 'thunder' in description_lower or 'storm' in description_lower:
            return '⛈️'
        elif 'snow' in description_lower:
            return '❄️'
        elif 'mist' in description_lower or 'fog' in description_lower:
            return '🌫️'
        else:
            return '🌤️'
    
    def get_outfit_suggestion(self, temperature: float) -> str:
        """Suggest outfit based on temperature"""
        if temperature < 0:
            return "Bundle up! Heavy coat, scarf, and gloves recommended. 🧥🧣"
        elif temperature < 10:
            return "Wear a warm jacket and layers. 🧥"
        elif temperature < 20:
            return "A light jacket or sweater should be good. 👔"
        elif temperature < 25:
            return "T-shirt and jeans weather! Perfect! 👕"
        elif temperature < 30:
            return "Light clothes recommended. Stay cool! 👕🩳"
        else:
            return "It's hot! Wear light, breathable clothes and stay hydrated! 🌡️💧"
    
    def get_activity_suggestion(self, weather_data: Dict) -> str:
        """Suggest activities based on weather"""
        description = weather_data['description'].lower()
        temp = weather_data['temperature']
        
        if 'rain' in description or 'storm' in description:
            return "Indoor day! Perfect for movies, reading, or coding. 🏠📚"
        elif 'clear' in description and 15 < temp < 25:
            return "Beautiful day! Great for a walk, picnic, or outdoor activities! 🌳🚶"
        elif temp > 30:
            return "Hot day! Swimming, air-conditioned mall, or stay indoors. 🏊"
        elif temp < 5:
            return "Cold! Hot chocolate and indoor activities recommended. ☕"
        else:
            return "Decent weather for outdoor activities if you dress appropriately! 🌤️"
    
    def get_smart_weather_response(self, city: str = None) -> str:
        """Generate intelligent weather response"""
        weather = self.get_weather(city)
        
        if not weather:
            if not self.api_key:
                return "I don't have a weather API key configured yet. You can get a free one from openweathermap.org!"
            return f"Sorry, I couldn't fetch the weather for {city or self.default_city} right now."
        
        emoji = self.get_weather_emoji(weather['description'])
        outfit = self.get_outfit_suggestion(weather['temperature'])
        activity = self.get_activity_suggestion(weather)
        
        response = f"""Weather in {weather['city']} {emoji}:
        
🌡️ Temperature: {weather['temperature']:.1f}°C (feels like {weather['feels_like']:.1f}°C)
💧 Humidity: {weather['humidity']}%
🌬️ Wind: {weather['wind_speed']} m/s
☁️ Conditions: {weather['description'].title()}

{outfit}

{activity}"""
        
        return response
    
    def is_good_weather(self, weather_data: Dict) -> bool:
        """Determine if weather is generally pleasant"""
        if not weather_data:
            return False
        
        description = weather_data['description'].lower()
        temp = weather_data['temperature']
        
        bad_conditions = ['rain', 'storm', 'snow', 'extreme']
        if any(cond in description for cond in bad_conditions):
            return False
        
        if 15 <= temp <= 25:
            return True
        
        return False

File: honey-rs/src/weather_service.rs
use anyhow::Result;
use serde_json::Value;
use std::time::Duration;

pub struct WeatherService {
    api_key: Option<String>,
    default_city: String,
    http: reqwest::blocking::Client,
}

pub struct WeatherData {
    pub city: String,
    pub temperature: f64,
    pub feels_like: f64,
    pub humidity: i64,
    pub description: String,
    pub wind_speed: f64,
}

impl WeatherService {
    pub fn new(api_key: Option<String>) -> Self {
        Self {
            api_key,
            default_city: crate::config::default_city(),
            http: reqwest::blocking::Client::builder()
                .timeout(Duration::from_secs(10))
                .build()
                .expect("http client"),
        }
    }

    pub fn get_weather(&self, city: Option<&str>) -> Result<WeatherData> {
        let Some(key) = &self.api_key else {
            anyhow::bail!("no weather API key configured");
        };
        let city = city.unwrap_or(&self.default_city);
        let resp = self
            .http
            .get("https://api.openweathermap.org/data/2.5/weather")
            .query(&[
                ("q", city),
                ("appid", key.as_str()),
                ("units", "metric"),
            ])
            .send()?;
        if !resp.status().is_success() {
            anyhow::bail!("weather service returned HTTP {}", resp.status());
        }
        let v: Value = resp.json()?;
        Ok(WeatherData {
            city: v["name"].as_str().unwrap_or(city).to_string(),
            temperature: v["main"]["temp"].as_f64().unwrap_or(0.0),
            feels_like: v["main"]["feels_like"].as_f64().unwrap_or(0.0),
            humidity: v["main"]["humidity"].as_i64().unwrap_or(0),
            description: v["weather"][0]["description"]
                .as_str()
                .unwrap_or("unknown")
                .to_string(),
            wind_speed: v["wind"]["speed"].as_f64().unwrap_or(0.0),
        })
    }

    pub fn smart_response(&self, city: Option<&str>) -> String {
        match self.get_weather(city) {
            Ok(w) => {
                let emoji = emoji_for(&w.description);
                let outfit = outfit_suggestion(w.temperature);
                let activity = activity_suggestion(&w.description, w.temperature);
                format!(
                    "Weather in {} {}: {:.1} degrees (feels like {:.1}), humidity {} percent, wind {:.1} meters per second, {}. {}. {}",
                    w.city, emoji, w.temperature, w.feels_like, w.humidity,
                    w.wind_speed, title_case(&w.description), outfit, activity
                )
            }
            Err(e) => {
                if self.api_key.is_none() {
                    "I don't have a weather API key configured yet. Set OWM_API_KEY and I'll check the sky for you!".into()
                } else {
                    format!("Sorry, I couldn't fetch the weather for {}. ({e})", city.unwrap_or(&self.default_city))
                }
            }
        }
    }
}

fn emoji_for(description: &str) -> &'static str {
    let d = description.to_lowercase();
    if d.contains("clear") {
        "\u{2600}\u{FE0F}"
    } else if d.contains("thunder") || d.contains("storm") {
        "\u{26A8}\u{FE0F}"
    } else if d.contains("snow") {
        "\u{2744}\u{FE0F}"
    } else if d.contains("rain") || d.contains("drizzle") {
        "\u{1F327}\u{FE0F}"
    } else if d.contains("cloud") {
        "\u{2601}\u{FE0F}"
    } else if d.contains("mist") || d.contains("fog") {
        "\u{1F32B}\u{FE0F}"
    } else {
        "\u{1F324}\u{FE0F}"
    }
}

fn outfit_suggestion(temp: f64) -> &'static str {
    if temp < 0.0 {
        "Bundle up! Heavy coat, scarf and gloves."
    } else if temp < 10.0 {
        "Wear a warm jacket and layers."
    } else if temp < 20.0 {
        "A light jacket or sweater should be good."
    } else if temp < 25.0 {
        "T-shirt and jeans weather!"
    } else if temp < 30.0 {
        "Light clothes recommended, stay cool."
    } else {
        "It's hot! Wear light breathable clothes and stay hydrated."
    }
}

fn activity_suggestion(description: &str, temp: f64) -> &'static str {
    let d = description.to_lowercase();
    if d.contains("rain") || d.contains("storm") {
        "Indoor day! Perfect for movies, reading or coding."
    } else if d.contains("clear") && (15.0..25.0).contains(&temp) {
        "Beautiful day! Great for a walk or a picnic."
    } else if temp > 30.0 {
        "Hot day! Swimming or staying in the AC sounds good."
    } else if temp < 5.0 {
        "Cold! Hot chocolate and indoor activities recommended."
    } else {
        "Decent weather for outdoor activities if you dress appropriately."
    }
}

fn title_case(s: &str) -> String {
    s.split_whitespace()
        .map(|w| {
            let mut c = w.chars();
            match c.next() {
                Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
                None => String::new(),
            }
        })
        .collect::<Vec<_>>()
        .join(" ")
}
